### PR TITLE
feat: highlight passages and tables on PDF in DocumentPreview

### PR DESCRIPTION
**Pull request by fukudasjp**
#### What do these changes do/fix?
Enable passage and table highlighting on PDF by integrating PDF highlighting with document preview.

This PR does the following changes:
- Use `PdfViewerWithHighlight` instead of `PdfViewer` in `DocumentPreview`
- Add capability of highlighting tables in `PdfViewerWithHighlight`
image
image

#### How do you test/verify these changes?
Use storybook to test the behavior:
- When you open the storybook, passage or table highlight is visible
- Toolbar shows the correct page number
- Page can move forward/backward using the toolbar

#### Are there any breaking changes included in this pull request?
no

### CODE REVIEW

File: packages/discovery-react-components/src/components/DocumentPreview/components/Highlight/TableHighlight.tsx [hunk 1]
```diff
@@ -54,18 +54,7 @@ export const TableHighlight: FC<Props> = ({
...
-    const loc = get(highlight, 'table.location');
-    if (!loc) {
-      return;
-    }
-
-    const { begin, end } = loc;
-    const table = processedDoc.tables.find(({ location }) =>
-      spansIntersect(location, { begin, end })
-    );
+    const table = getHighlightTable(highlight, processedDoc);
```
Comment (fukudasjp) [inline packages/discovery-react-components/src/components/DocumentPreview/components/Highlight/TableHighlight.tsx:57]: Extracted a method to `tables.ts`

File: packages/discovery-react-components/src/components/DocumentPreview/components/PdfHighlight/types.ts [hunk 2]
```diff
@@ -55,12 +67,12 @@ export interface HighlightProps {
...
    * Document data returned by query
    */
-  document: QueryResult;
+  document?: QueryResult;
 
   /**
    * Highlight spans on fields in document
    */
-  highlights: DocumentFieldHighlight[];
+  highlights?: DocumentFieldHighlight[];
```
Comment (fukudasjp) [inline packages/discovery-react-components/src/components/DocumentPreview/components/PdfHighlight/types.ts:75]: Make these props optional so that the components can work for Table highlight.

File: packages/discovery-react-components/src/components/DocumentPreview/DocumentPreview.tsx [hunk 3]
```diff
@@ -163,20 +164,23 @@ function PreviewDocument({
...
 }: PreviewDocumentProps): ReactElement | null {
   // if we have PDF data, render that
   // otherwise, render fallback document view
   if (file) {
     return (
-      <PdfViewer
+      <PdfViewerWithHighlight
         file={file}
-        document={document}
+        document={document ?? undefined}
```
Comment (jhpedemonte) [inline packages/discovery-react-components/src/components/DocumentPreview/DocumentPreview.tsx:176]: Why is this necessary?

Comment (jhpedemonte) [reply packages/discovery-react-components/src/components/DocumentPreview/DocumentPreview.tsx:176]: `Props` above defines it as:

```
document?: QueryResult;
```

which is equivalent to `QueryResult || undefined`.

File: packages/discovery-react-components/src/components/DocumentPreview/components/Highlight/tables.tsx [hunk 4]
```diff
@@ -0,0 +1,24 @@
+import { QueryTableResult } from 'ibm-watson/discovery/v2';
+import { ProcessedDoc } from 'utils/document';
+import { spansIntersect } from 'utils/document/documentUtils';
+
+export function isTable(highlight: any): highlight is QueryTableResult {
+  return !!highlight?.table?.location;
+}
+
+export function getHighlightTable(
```
Comment (jhpedemonte) [inline packages/discovery-react-components/src/components/DocumentPreview/components/Highlight/tables.tsx:9]: How about `getHighlightedTable`? Just a little better name.

File: packages/discovery-react-components/src/components/DocumentPreview/components/Highlight/__tests__/tables.test.tsx [hunk 5]
```diff
@@ -0,0 +1,71 @@
+import { QueryResultPassage, QueryTableResult } from 'ibm-watson/discovery/v2';
```
Comment (jhpedemonte) [inline packages/discovery-react-components/src/components/DocumentPreview/components/Highlight/__tests__/tables.test.tsx:1]: This file can just be named `.ts`, since it's not rendering any JSX.

File: packages/discovery-react-components/src/components/DocumentPreview/components/Highlight/passages.tsx [hunk 6]
```diff
@@ -57,6 +57,6 @@ export function getPassagePageInfo(
 }
 
 // eslint-disable-next-line @typescript-eslint/no-explicit-any
-export function isPassage(obj: any): boolean {
+export function isPassage(obj: any): obj is QueryResultPassage {
```
Comment (jhpedemonte) [inline packages/discovery-react-components/src/components/DocumentPreview/components/Highlight/passages.tsx:60]: I don't understand this change. Is it correct?

Should this be this instead, and we can remove the `eslint` comment?

```js
export function isPassage(obj: QueryResultPassage): boolean {
```

File: packages/discovery-react-components/src/components/DocumentPreview/components/Highlight/tables.tsx [hunk 7]
```diff
@@ -0,0 +1,24 @@
+import { QueryTableResult } from 'ibm-watson/discovery/v2';
```
Comment (jhpedemonte) [inline packages/discovery-react-components/src/components/DocumentPreview/components/Highlight/tables.tsx:1]: Also just name `.ts`.

File: packages/discovery-react-components/src/components/DocumentPreview/components/Highlight/tables.tsx [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+import { ProcessedDoc } from 'utils/document';
+import { spansIntersect } from 'utils/document/documentUtils';
+
+export function isTable(highlight: any): highlight is QueryTableResult {
```
Comment (jhpedemonte) [inline packages/discovery-react-components/src/components/DocumentPreview/components/Highlight/tables.tsx:5]: Same here, I don't understand the `highlight is QueryTableResult`.

**Review by jhpedemonte — APPROVED** [review]
Looks good, both functionality and code. I have a few small nits and questions, so I'm going to approve, so you're not blocked.

File: packages/discovery-react-components/src/components/DocumentPreview/components/Highlight/passages.tsx (hunk 6 above)
Comment (fukudasjp) [reply packages/discovery-react-components/src/components/DocumentPreview/components/Highlight/passages.tsx:60]: Yes, I wanted to introduce [type predicate](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates) for TS compiler to assume the `obj` is `QueryResultPassage` when the function returns `true`. Using this, we can avoid (error-prone) type annotations in caller.

I just realized that I haven't updated the callers... I'll do that.

File: packages/discovery-react-components/src/components/DocumentPreview/DocumentPreview.tsx (hunk 3 above)
Comment (fukudasjp) [reply packages/discovery-react-components/src/components/DocumentPreview/DocumentPreview.tsx:176]: It was necessary because the `document` there is `QueryResult |  null | undefined`. I revise the type definition as I agree that `document ?? undefined` looks strange.

File: packages/discovery-react-components/src/components/DocumentPreview/components/Highlight/tables.tsx (hunk 7 above)
Comment (fukudasjp) [reply packages/discovery-react-components/src/components/DocumentPreview/components/Highlight/tables.tsx:1]: good catch! I rename `passages.tsx` as well :)
